File: app/src/lib/downloadPattern.ts
export const DEFAULT_DOWNLOAD_PATTERN = "inkbunny/{artist}/{file_name_full}"

export type DownloadPatternToken = {
  name: string
  label: string
  description: string
  example: string
}

export type DownloadPatternSegment =
  | { kind: "text"; value: string }
  | { kind: "token"; value: string; token: DownloadPatternToken }
  | { kind: "invalid"; value: string; name: string }

const TOKEN_MAP = new Map<string, DownloadPatternToken>()

export const DOWNLOAD_PATTERN_TOKENS: DownloadPatternToken[] = [
  {
    name: "artist",
    label: "Artist",
    description: "Submission artist username.",
    example: "elly",
  },
  {
    name: "artist_id",
    label: "Artist ID",
    description: "Numeric artist account ID.",
    example: "12345",
  },
  {
    name: "title",
    label: "Title",
    description: "Submission title.",
    example: "Star Patrol",
  },
  {
    name: "rating",
    label: "Rating",
    description: "Coarse rating bucket, prioritized as adult, mature, then general.",
    example: "adult",
  },
  {
    name: "public",
    label: "Visibility",
    description: "Submission visibility as public or private.",
    example: "public",
  },
  {
    name: "submission_type",
    label: "Type",
    description: "Sanitized one-word submission type such as picture, series, video, or music.",
    example: "series",
  },
  {
    name: "year",
    label: "Year",
    description: "Submission year from the upload timestamp.",
    example: "2026",
  },
  {
    name: "month",
    label: "Month",
    description: "Submission month using two digits.",
    example: "03",
  },
  {
    name: "day",
    label: "Day",
    description: "Submission day using two digits.",
    example: "09",
  },
  {
    name: "hour",
    label: "Hour",
    description: "Submission hour using 24-hour time.",
    example: "16",
  },
  {
    name: "minute",
    label: "Minute",
    description: "Submission minute using two digits.",
    example: "27",
  },
  {
    name: "file_name_full",
    label: "Full filename",
    description: "Original file name including its extension.",
    example: "27491_elly_star-patrol.png",
  },
  {
    name: "file_name",
    label: "Filename stem",
    description: "Original file name without the extension.",
    example: "27491_elly_star-patrol",
  },
  {
    name: "file_name_ext",
    label: "Filename stem",
    description: "Alias for the file name without the extension.",
    example: "27491_elly_star-patrol",
  },
  {
    name: "file_id",
    label: "File ID",
    description: "Numeric file ID.",
    example: "27491",
  },
  {
    name: "number",
    label: "File number",
    description: "1-based file order inside the submission.",
    example: "3",
  },
  {
    name: "ext",
    label: "Ext",
    description: "File extension without the dot.",
    example: "png",
  },
  {
    name: "extension",
    label: "Extension",
    description: "Alias for the extension without the dot.",
    example: "png",
  },
  {
    name: "submission_id",
    label: "Submission ID",
    description: "Numeric submission ID.",
    example: "908172",
  },
  {
    name: "pool_id",
    label: "Pool ID",
    description:
      "Pool ID. Using pool tokens duplicates files into one folder per pool, and the folder segment is omitted when a submission has no pools.",
    example: "7001",
  },
  {
    name: "pool_name",
    label: "Pool name",
    description:
      "Pool name. Using pool tokens duplicates files into one folder per pool, and the folder segment is omitted when a submission has no pools.",
    example: "season-one",
  },
]

for (const token of DOWNLOAD_PATTERN_TOKENS) {
  TOKEN_MAP.set(token.name, token)
}

const TOKEN_RE = /\{([a-z0-9_]+)\}/g
const PREVIEW_POOLS = [
  { pool_id: "7001", pool_name: "season-one" },
  { pool_id: "7002", pool_name: "favorites" },
]

const PREVIEW_VALUES: Record<string, string> = {
  artist: "elly",
  artist_id: "12345",
  title: "Star Patrol",
  rating: "adult",
  public: "public",
  submission_type: "series",
  year: "2026",
  month: "03",
  day: "09",
  hour: "16",
  minute: "27",
  file_name_full: "27491_elly_star-patrol.png",
  file_name: "27491_elly_star-patrol",
  file_name_ext: "27491_elly_star-patrol",
  file_id: "27491",
  number: "3",
  ext: "png",
  extension: "png",
  submission_id: "908172",
}

export function tokenizeDownloadPattern(value: string): DownloadPatternSegment[] {
  const segments: DownloadPatternSegment[] = []
  let lastIndex = 0

  value.replaceAll(TOKEN_RE, (match, name: string, offset: number) => {
    if (offset > lastIndex) {
      segments.push({ kind: "text", value: value.slice(lastIndex, offset) })
    }

    const token = TOKEN_MAP.get(name)
    if (token) {
      segments.push({ kind: "token", value: match, token })
    } else {
      segments.push({ kind: "invalid", value: match, name })
    }

    lastIndex = offset + match.length
    return match
  })

  if (lastIndex < value.length) {
    segments.push({ kind: "text", value: value.slice(lastIndex) })
  }

  return segments
}

export function collectUnknownDownloadTokens(value: string): string[] {
  const unknown = new Set<string>()

  value.replaceAll(TOKEN_RE, (match, name: string) => {
    if (!TOKEN_MAP.has(name)) {
      unknown.add(match)
    }
    return match
  })

  return [...unknown]
}

export function renderDownloadPatternPreview(pattern: string): string[] {
  const activePattern = pattern.trim() || DEFAULT_DOWNLOAD_PATTERN
  const normalized = activePattern.replaceAll("\\", "/")
  const usesPoolTokens = normalized.includes("{pool_id}") || normalized.includes("{pool_name}")
  const contexts = usesPoolTokens ? PREVIEW_POOLS : [{}]

  return contexts.map((poolValues) => {
    const poolRecord = poolValues as Record<string, string>
    const rendered = normalized.replaceAll(TOKEN_RE, (match, name: string) => {
      if (name in poolRecord) {
        return poolRecord[name] ?? ""
      }
      return PREVIEW_VALUES[name] ?? match
    })
    return normalizePreviewPath(rendered)
  })
}

function normalizePreviewPath(value: string): string {
  const segments = value
    .split(/[\\/]/)
    .map((segment) => sanitizePathComponent(segment))
    .filter(Boolean)

  if (segments.length === 0) {
    return DEFAULT_DOWNLOAD_PATTERN
  }

  return segments.join("/")
}

function sanitizePathComponent(value: string): string {
  return value
    .trim()
    .replace(/[<>:"|?*\u0000-\u001f]/g, "")
    .replace(/[\\/]/g, "_")
    .replace(/[. ]+$/g, "")
}

File: pkg/app/download_pattern.go
package desktopapp

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
	"unicode"

	"github.com/ellypaws/inkbunny"
)

const defaultDownloadPattern = "inkbunny/{artist}/{file_name_full}"

var downloadPatternTokenRE = regexp.MustCompile(`\{([a-z0-9_]+)\}`)

type downloadPathContext struct {
	Submission inkbunny.SubmissionDetails
	File       inkbunny.File
	Pool       *inkbunny.Pool
	Time       time.Time
	Number     int
}

func normalizeDownloadPattern(pattern string) string {
	trimmed := strings.TrimSpace(pattern)
	if trimmed == "" {
		return defaultDownloadPattern
	}
	return trimmed
}

func resolveDownloadDestinations(
	root string,
	pattern string,
	submission inkbunny.SubmissionDetails,
	file inkbunny.File,
) []string {
	cleanRoot := filepath.Clean(strings.TrimSpace(root))
	if cleanRoot == "" {
		return nil
	}

	pattern = normalizeDownloadPattern(pattern)
	baseContext := downloadPathContext{
		Submission: submission,
		File:       file,
		Time:       resolveDownloadTimestamp(submission, file),
		Number:     resolveFileNumber(file),
	}

	if !patternUsesPoolTokens(pattern) || len(submission.Pools) == 0 {
		return []string{renderDownloadDestination(cleanRoot, pattern, baseContext)}
	}

	destinations := make([]string, 0, len(submission.Pools))
	for i := range submission.Pools {
		pool := submission.Pools[i]
		ctx := baseContext
		ctx.Pool = &pool
		destinations = append(destinations, renderDownloadDestination(cleanRoot, pattern, ctx))
	}
	return uniqueNonEmptyPaths(destinations)
}

func renderDownloadDestination(root, pattern string, ctx downloadPathContext) string {
	normalizedPattern := strings.ReplaceAll(normalizeDownloadPattern(pattern), "\\", "/")
	rawParts := strings.Split(normalizedPattern, "/")
	segments := make([]string, 0, len(rawParts)+1)
	lastPartHasValue := false

	for i, rawPart := range rawParts {
		rendered := sanitizePathComponent(renderDownloadComponent(rawPart, ctx))
		if i == len(rawParts)-1 && rendered != "" {
			lastPartHasValue = true
		}
		if rendered == "" {
			continue
		}
		segments = append(segments, rendered)
	}

	if !lastPartHasValue {
		fallback := sanitizePathComponent(filepath.Base(ctx.File.FileName))
		if fallback == "" {
			fallback = fmt.Sprintf("file_%s", ctx.File.FileID.String())
		}
		segments = append(segments, fallback)
	}

	if len(segments) == 0 {
		segments = []string{
			"inkbunny",
			sanitizePathComponent(filepath.Base(ctx.File.FileName)),
		}
	}

	return filepath.Clean(filepath.Join(root, filepath.Join(segments...)))
}

func renderDownloadComponent(component string, ctx downloadPathContext) string {
	return downloadPatternTokenRE.ReplaceAllStringFunc(component, func(match string) string {
		name := downloadPatternTokenRE.FindStringSubmatch(match)
		if len(name) != 2 {
			return match
		}

		value, ok := downloadTokenValue(name[1], ctx)
		if !ok {
			return match
		}
		return value
	})
}

func downloadTokenValue(name string, ctx downloadPathContext) (string, bool) {
	fileBase := filepath.Base(ctx.File.FileName)
	fileStem := strings.TrimSuffix(fileBase, filepath.Ext(fileBase))
	ext := strings.TrimPrefix(strings.ToLower(filepath.Ext(fileBase)), ".")

	switch name {
	case "artist":
		return ctx.Submission.Username, true
	case "artist_id":
		return ctx.Submission.UserID.String(), true
	case "title":
		return ctx.Submission.Title, true
	case "rating":
		return coarseSubmissionRating(ctx.Submission), true
	case "public":
		return ternary(ctx.Submission.Public.Bool(), "public", "private"), true
	case "submission_type":
		return normalizedSubmissionType(ctx.Submission.SubmissionTypeID.Int()), true
	case "year":
		return formatTimePart(ctx.Time, "2006"), true
	case "month":
		return formatTimePart(ctx.Time, "01"), true
	case "day":
		return formatTimePart(ctx.Time, "02"), true
	case "hour":
		return formatTimePart(ctx.Time, "15"), true
	case "minute":
		return formatTimePart(ctx.Time, "04"), true
	case "file_name_full":
		return fileBase, true
	case "file_name":
		return fileStem, true
	case "file_name_ext":
		return fileStem, true
	case "file_id":
		return ctx.File.FileID.String(), true
	case "number":
		return fmt.Sprintf("%d", ctx.Number), true
	case "ext":
		return ext, true
	case "extension":
		return ext, true
	case "submission_id":
		return ctx.Submission.SubmissionID.String(), true
	case "pool_id":
		if ctx.Pool == nil {
			return "", true
		}
		return ctx.Pool.PoolID.String(), true
	case "pool_name":
		if ctx.Pool == nil {
			return "", true
		}
		return ctx.Pool.Name, true
	default:
		return "", false
	}
}

func downloadTargetsMatch(destinations []string, expectedMD5 string) (bool, int64, string, error) {
	unique := uniqueNonEmptyPaths(destinations)
	if len(unique) == 0 {
		return false, 0, "", nil
	}

	allMatch := true
	var size int64
	source := ""
	for _, destination := range unique {
		result, err := verifyDownloadedFile(destination, expectedMD5)
		if err != nil {
			return false, 0, "", err
		}
		if result.Matches {
			if source == "" {
				source = destination
				size = result.Size
			}
			continue
		}
		allMatch = false
	}
	return allMatch, size, source, nil
}

func ensureDownloadTargetsFromSource(source string, destinations []string, expectedMD5 string) error {
	if strings.TrimSpace(source) == "" {
		return nil
	}

	cleanSource := filepath.Clean(source)
	for _, destination := range uniqueNonEmptyPaths(destinations) {
		cleanDestination := filepath.Clean(destination)
		if cleanDestination == cleanSource {
			continue
		}

		result, err := verifyDownloadedFile(cleanDestination, expectedMD5)
		if err != nil {
			return err
		}
		if result.Matches {
			continue
		}
		if result.Exists {
			if err := os.Remove(cleanDestination); err != nil && !os.IsNotExist(err) {
				return err
			}
		}
		if err := os.MkdirAll(filepath.Dir(cleanDestination), 0o755); err != nil {
			return err
		}
		if err := copyFile(cleanSource, cleanDestination); err != nil {
			return err
		}
	}

	return nil
}

func writeKeywordSidecars(destinations []string, keywords string) error {
	if strings.TrimSpace(keywords) == "" {
		return nil
	}
	for _, destination := range uniqueNonEmptyPaths(destinations) {
		sidecar := stringsTrimExt(destination) + ".txt"
		if err := os.MkdirAll(filepath.Dir(sidecar), 0o755); err != nil {
			return err
		}
		if err := os.WriteFile(sidecar, []byte(keywords), 0o600); err != nil {
			return err
		}
	}
	return nil
}

func copyFile(source, destination string) error {
	in, err := os.Open(source)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.Create(destination)
	if err != nil {
		return err
	}
	defer func() {
		_ = out.Close()
	}()

	if _, err := io.Copy(out, in); err != nil {
		return err
	}
	return out.Sync()
}

func uniqueNonEmptyPaths(paths []string) []string {
	seen := make(map[string]struct{}, len(paths))
	unique := make([]string, 0, len(paths))
	for _, path := range paths {
		clean := filepath.Clean(strings.TrimSpace(path))
		if clean == "." || clean == "" {
			continue
		}
		if _, ok := seen[clean]; ok {
			continue
		}
		seen[clean] = struct{}{}
		unique = append(unique, clean)
	}
	return unique
}

func sanitizePathComponent(value string) string {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return ""
	}

	cleaned := strings.Map(func(r rune) rune {
		switch {
		case r == '/' || r == '\\':
			return '_'
		case r == '<' || r == '>' || r == ':' || r == '"' || r == '|' || r == '?' || r == '*':
			return -1
		case unicode.IsControl(r):
			return -1
		default:
			return r
		}
	}, trimmed)

	cleaned = strings.Trim(cleaned, " .")
	if cleaned == "" {
		return ""
	}
	return cleaned
}

func patternUsesPoolTokens(pattern string) bool {
	return strings.Contains(pattern, "{pool_id}") || strings.Contains(pattern, "{pool_name}")
}

func resolveDownloadTimestamp(submission inkbunny.SubmissionDetails, file inkbunny.File) time.Time {
	for _, candidate := range []string{
		submission.CreateDateUser,
		submission.CreateDateSystem,
		file.CreateDateTimeUser,
		file.CreateDateTime,
	} {
		if parsed, ok := parseDownloadTime(candidate); ok {
			return parsed
		}
	}
	return time.Time{}
}

func parseDownloadTime(value string) (time.Time, bool) {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return time.Time{}, false
	}
	layouts := []string{
		"2006-01-02 15:04:05",
		"2006-01-02T15:04:05",
		time.RFC3339,
	}
	for _, layout := range layouts {
		parsed, err := time.Parse(layout, trimmed)
		if err == nil {
			return parsed, true
		}
	}
	return time.Time{}, false
}

func formatTimePart(value time.Time, layout string) string {
	if value.IsZero() {
		return ""
	}
	return value.Format(layout)
}

func resolveFileNumber(file inkbunny.File) int {
	number := file.SubmissionFileOrder.Int() + 1
	if number <= 0 {
		return 1
	}
	return number
}

func coarseSubmissionRating(submission inkbunny.SubmissionDetails) string {
	name := strings.ToLower(strings.TrimSpace(submission.RatingName))
	if strings.Contains(name, "adult") || strings.Contains(name, "sexual") {
		return "adult"
	}
	if strings.Contains(name, "mature") || strings.Contains(name, "nudity") || strings.Contains(name, "violence") {
		return "mature"
	}
	for _, rating := range submission.Ratings {
		tag := strings.ToLower(strings.TrimSpace(rating.Name))
		if strings.Contains(tag, "sexual") {
			return "adult"
		}
		if strings.Contains(tag, "nudity") || strings.Contains(tag, "violence") {
			return "mature"
		}
	}
	return "general"
}

func normalizedSubmissionType(typeID int) string {
	switch typeID {
	case int(inkbunny.SubmissionTypePicturePinup):
		return "picture"
	case int(inkbunny.SubmissionTypeSketch):
		return "sketch"
	case int(inkbunny.SubmissionTypePictureSeries):
		return "series"
	case int(inkbunny.SubmissionTypeComic):
		return "comic"
	case int(inkbunny.SubmissionTypePortfolio):
		return "portfolio"
	case int(inkbunny.SubmissionTypeShockwaveFlashAnimation),
		int(inkbunny.SubmissionTypeShockwaveFlashInteractive),
		int(inkbunny.SubmissionTypeVideoFeatureLength),
		int(inkbunny.SubmissionTypeVideoAnimation3DCGI):
		return "video"
	case int(inkbunny.SubmissionTypeMusicSingleTrack),
		int(inkbunny.SubmissionTypeMusicAlbum):
		return "music"
	case int(inkbunny.SubmissionTypeWritingDocument):
		return "stories"
	case int(inkbunny.SubmissionTypeCharacterSheet):
		return "character_sheet"
	case int(inkbunny.SubmissionTypePhotography):
		return "photography"
	default:
		return "submission"
	}
}
